src/cache/BrowserBasedPackageCache.ts

```diff
@@ -0,0 +1,223 @@
+    }
+    const resource = packageMap.get(resourceLabel);
+    if (resource == null) {
+      throw new Error(`Failed to get resource at path ${resourcePath}`);
+    }
+    return resource;
+  }
+}
+
+async function addToIndexedDB(databaseName: string, packageLabel: string, resources: any[]) {
+  const databaseVersion = await getNextDatabaseVersion(databaseName);
+  return addResourcesToDatabase(databaseName, databaseVersion, packageLabel, resources);
+}
+
+async function getNextDatabaseVersion(databaseName: string): Promise<number> {
+  return new Promise((resolve, reject) => {
+    const openRequest = indexedDB.open(databaseName);
+    openRequest.onsuccess = event => {
+      const database = (event.target as IDBOpenDBRequest).result;
+      const nextVersion = database.version + 1;
+      database.close();
+      resolve(nextVersion);
+    };
+    openRequest.onerror = error => reject(error);
+  });
+}
+
+async function addResourcesToDatabase(
+  databaseName: string,
+  databaseVersion: number,
+  packageLabel: string,
+  resources: any[]
+): Promise<any[]> {
+  const savedResources: any[] = [];
+  return new Promise((resolve, reject) => {
+    const openRequest = indexedDB.open(databaseName, databaseVersion);
+    openRequest.onupgradeneeded = event => {
+      const database = (event.target as IDBOpenDBRequest).result;
+      database.createObjectStore(packageLabel, { keyPath: ['id', 'resourceType'] });
+    };
+    openRequest.onsuccess = event => {
+      const database = (event.target as IDBOpenDBRequest).result;
+      const transaction = database.transaction([packageLabel], 'readwrite');
+      const objectStore = transaction.objectStore(packageLabel);
+      resources.forEach(resource => {
+        if (resource.id && resource.resourceType) {
+          objectStore.put(resource);
+          savedResources.push(resource);
+        }
+      });
+      transaction.oncomplete = () => {
+        database.close();
+        resolve(savedResources);
+      };
+      transaction.onerror = () => {
+        database.close();
+        reject('Unexpected error adding package to cache');
+      };
+    };
+    openRequest.onerror = () => {
+      reject('Unexpected error saving package to cache');
+    };
+  });
+}
+
+function initializeResourcesInLocalMap(
+  database: any,
+  packageLabel: string
+): Promise<[string, Map<string, any>]> {
+  return new Promise((resolve, reject) => {
+    const transaction = database.transaction([packageLabel]);
+    const objectStore = transaction.objectStore(packageLabel);
+    const request = objectStore.getAll();
+    request.onerror = () => {
+      reject('Unexpected error trying to initialize');
+    };
+    request.onsuccess = (event: any) => {
+      const listOfResources: [string, any][] = event.target.result.map((r: any) => [
+        `${r.resourceType}-${r.id}`,
+        r
+      ]);
+      const resourceMap = new Map(listOfResources);
```

Comment:
Is there a reason to do it as above, vs:
```ts
const resourceMap = new Map();
event.target.result.forEach((r: any) => resourceMap.set(`${r.resourceType}-${r.id}`, r));
```
There might be. I'm hoping to learn something!

Reply:
Uhh no I don't believe there's any reason to do it that way! I was probably looking at what was the most efficient way to add everything to the map, but I think either of the above ways is the same. Alternatively, I actually realized earlier in the code I did something similar this way:

```javascript
const resourceMap= new Map(event.target.result.map((r: any) => [`${r.resourceType}-${r.id}`, r]));
```

I think as far as I can tell, all three of these are the same performance-wise and either your suggestion or this latest change are the easiest to understand. I went with this one for consistency, but happy to change them both to use the `.set()` within a `forEach` if you think it is preferred.